### PR TITLE
Add `from_preset` constructor to `BertPreprocessor`

### LINKED ISSUE #388: Add `from_preset` constructor to `BertPreprocessor`
# Proposal

In #387 we allowed construction of a BERT model from a "preset" model architecture and weights; for example `Bert.from_preset("bert_base_uncased_en")`. I propose to do the same with `BertPreprocessor`, automatically generating the matching preprocessing functionality using the same string id: `BertPreprocessor.from_preset("bert_base_uncased_en")`.

# Preset configuration

We will need to add preprocessing metadata to each preset id. One way to do this is keeping the vocabulary information normalized in a separate struct and joining it at call time. However, we could follow the pattern in #387 more closely by inlining the preprocessing config and weights:

```python
backbone_presets = {
    "bert_tiny_uncased_en": {
        "config": {
            "vocabulary_size": 30522,
            "num_layers": 2,
            "num_heads": 2,
            "hidden_dim": 128,
            "intermediate_dim": 512,
            "dropout": 0.1,
            "max_sequence_length": 512,
            "num_segments": 2,
        },
        "description": (
            "Tiny size of BERT where all input is lowercased. "
            "Trained on English Wikipedia + BooksCorpus."
        ),
        "weights_url": "https://storage.googleapis.com/keras-nlp/models/bert_tiny_uncased_en/model.h5",
        "weights_hash": "c2b29fcbf8f814a0812e4ab89ef5c068",
        "vocabulary_url": "https://storage.googleapis.com/keras-nlp/models/bert_base_uncased_en/vocab.txt",
        "vocabulary_hash": "64800d5d8528ce344256daf115d4965e",
        "lowercase": True,
    },
}
```

If we add more preprocessor arguments than `lowercase` in the future then a separate `preprocessor_config` field could be added. However, we have already hardcoded most preprocessing metadata in the `BertPreprocessor` class. Additional information is generally added in files like `vocab.txt` and `merges.txt` rather than flat parameters, so this doesn't seem helpful at this time.

# Signature
The arguments `sequence_length` and `truncate` are not determined by the preset alone and should be added to the signature. We must enforce that `sequence_length` is no longer than the training time setting however.

One complication is that most users will want to use the biggest sequence length that the model allows. We might want to set the default to `None` so that we can fill in this value unless the user has a different preference.

```python
class BertPreprocessor(keras.layers.Layer):
    # Generic constructor
    def __init__(
        self,
        vocabulary="uncased_en",
        lowercase=False,
        sequence_length=512,
        truncate="round_robin",
        **kwargs,
    ):
        pass

    @classmethod
    def from_preset(
        cls, 
        preset,
        sequence_length=512, 
        truncate="round_robin",
        **kwargs,
    ):
    if sequence_length > backbone_presets[preset]["config"]["max_sequence_length"]:
        raise ValueError(...)
```

# Preset namespaces
In the long term we will likely add task specific presets as well as backbone presets. Unlike model classes, Preprocessors could pertain to any preset type. Therefore long term we will have to create a single namespace of presets or explicitly check multiple namespaces and enforce uniqueness between them.

### PR DESCRIPTION
**Pull request by jbischof**
Closes #388

### CODE REVIEW

File: keras_nlp/models/bert/bert_preprocessing.py [hunk 1]
```diff
@@ -18,119 +18,105 @@
...
 from keras_nlp.layers.multi_segment_packer import MultiSegmentPacker
-from keras_nlp.models.bert.bert_presets import vocabularies
+from keras_nlp.models.bert.bert_models import Bert
 from keras_nlp.tokenizers.word_piece_tokenizer import WordPieceTokenizer
 
 
-def _handle_pretrained_tokenizer_arguments(vocabulary, lowercase):
-    """Look up pretrained defaults for tokenizer arguments.
+def download_vocabulary(preset, url, hash):
+    """Download vocabulary pretaining to preset."""
```
Comment (fchollet) [inline keras_nlp/models/bert/bert_preprocessing.py:26]: "pertaining" (but you might want to use "associated with" to increase Levenshtein distance from "pretrained")

File: keras_nlp/models/bert/bert_presets.py [hunk 2]
```diff
@@ -45,13 +49,15 @@
...
-        "vocabulary": "uncased_en",
         "description": (
             "Small size of BERT where all input is lowercased. "
             "Trained on English Wikipedia + BooksCorpus."
         ),
         "weights_url": "https://storage.googleapis.com/keras-nlp/models/bert_small_uncased_en/model.h5",
         "weights_hash": "08632c9479b034f342ba2c2b7afba5f7",
+        "vocabulary_url": "https://storage.googleapis.com/keras-nlp/models/bert_base_uncased_en/vocab.txt",
+        "vocabulary_hash": "64800d5d8528ce344256daf115d4965e",
+        "lowercase": True,
```
Comment (fchollet) [inline keras_nlp/models/bert/bert_presets.py:60]: Perhaps there should be a separate tokenizer config field in the dict, to store things like "lowercase"?

**Review by fchollet — COMMENTED** [review]
Design Q: should preprocessors solely be available via

```
Preprocessor.from_preset(name)
```

or should they be *also* be available from the corresponding `name` model, e.g.

```
Bert.from_preset(name).tokenizer # or something
```

File: keras_nlp/models/bert/bert_presets.py (hunk 2 above)
Comment (mattdangerw) [reply keras_nlp/models/bert/bert_presets.py:60]: I like a config dict for this overall as well. I think it is a more general purpose and allows to grow a bit better. 

Even for sentencepiece, which is a "vocabulary" file format that stores a lot of it's own metadata, I see a need for this coming up. If you look at say the [t5 config](https://huggingface.co/t5-base/blob/main/config.json) in huggingface, they have a lot of preprocessing params that flatten straight in to the base config, including special token ID numbers, sequence lengths, even tasks specific prefixes.

I think collocating preprocessing config params in a dict gives us more room to grow.

**Comment by jbischof** [comment]
@fchollet, @mattdangerw is working on a `pipeline` class that composes preprocessing and the model. It would have its own `from_preset` method. Looking forward to discussing when we have a design doc.

**Comment by fchollet** [comment]
> @fchollet, @mattdangerw is working on a pipeline class that composes preprocessing and the model

I am not sure that a separate "pipeline" class is warranted. I'd favor including the preprocessing in the user-facing model (e.g. Bert or BertClassifier), similar to what we do for vision models.

- Models are instantiated with or without preprocessing included (defaults to True, i.e. with).
- With `fit()`, `predict()`, and `evaluate()`: the preprocessing is automatically mapped into the dataset (if it is included in the model).
- With `__call__()`: preprocessing is applied by default (if it is included in the model) but you have an argument to turn it off, which can be useful in various situations when writing low-level code.

**Comment by jbischof** [comment]
@fchollet I was assuming that regardless of how we implement preprocessing as part of the model we still need separate preprocessing to be a good experience, which I have attempted in this PR. Are you saying you want to block this PR until we figure out joint preprocessing or that the current PR is a bad experience for separate preprocessing?

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Overall this looks good to me! And definitely a positive delta from what we have.

I think one bigger question is whether we want to have a `BertPreprocessor.presets`. Soon, `Bert.from_preset`, `BertClassifier.from_preset` and `BertPreprocessor.from_preset` will all have different sets of keys they accept (the latter being the combination of the two former).

Should we try to keep a consistent UX, where each class has a `preset` static property, and the valid keys can be accessed with `MySymbol.preset.keys()`? I don't see a reason we need to figure that out to submit this, and there is probably plenty to discuss in the weeds there. But something we should think about going forward.

**Comment by mattdangerw** [comment]
@fchollet and @jbischof, re how to access the pre-preprocessing paired with a model, I think we can do that as a follow up. (Basically I'm just echoing Jonathan's point above).

IMO, we should have a way to access preprocessing standalone--that's a totally valid user journey. This PR address that in a very consistent way with the UX we just came up with for `keras_nlp.models.Bert`.

As Francois was saying in the original question, the question is whether to *also* expose the preprocessing on the model. I think that is a fairly dense topic, and one that can roughly be summarized as "what do we want to do about pipelines"? Let's tackle that in a future PR!

**Comment by fchollet** [comment]
> Are you saying you want to block this PR until we figure out joint preprocessing or that the current PR is a bad experience for separate preprocessing?

No, it's clear that being able to retrieve the tokenizer independently is useful. I was saying:

> should they be **also** be available from the corresponding name model

File: keras_nlp/models/bert/bert_preprocessing.py [hunk 3]
```diff
@@ -185,9 +171,88 @@ def call(self, inputs):
...
+
+        vocabulary = _download_vocabulary(
+            preset,
+            metadata["vocabulary_url"],
+            metadata["vocabulary_hash"],
+        )
+        config = metadata["preprocessor_config"]
+        # Use model's `max_sequence_length` if `sequence_length` unspecified;
+        # otherwise check that `sequence_length` not too long.
+        max_sequence_length = metadata["config"]["max_sequence_length"]
```
[the remark this answers is not in the document]
Comment (jbischof) [reply keras_nlp/models/bert/bert_preprocessing.py:198]: Adding a `presets` property to every class of compatible ids sounds like a reasonable idea. Done for this class!

Re: config, there is a lot of work behind the scenes to get the inputs for the generic constructor. I don't think it can be very clean with `vocabulary` as an argument.

File: keras_nlp/models/bert/bert_presets.py [hunk 4]
```diff
@@ -14,6 +14,8 @@
 
 # TODO(jbischof): document presets in keras.io and use URL in docstrings
 # Metadata for loading pretrained model weights.
+# TODO(mattdangerw): Update our bucket structure so the vocabularies are
```
[the remark this answers is not in the document]
Comment (jbischof) [reply keras_nlp/models/bert/bert_presets.py:17]: Yup, but it's not done yet.

**Comment by chenmoneygithub** [comment]
I tried this implementation and spotted a bug:
```
bert_preprocessor = keras_nlp.models.BertPreprocessor.from_preset("bert_small_uncased_en")
inputs = ["This is a testing string!", "Another test string!"]
inputs = bert_preprocessor(inputs)
print(inputs)
```

This gives shape (512, ) for `token_ids`, `segment_ids` and `padding_mask`, which means the batch information is lost.

File: keras_nlp/models/bert/bert_preprocessing.py [hunk 5]
```diff
@@ -18,119 +18,105 @@
...
 
 from keras_nlp.layers.multi_segment_packer import MultiSegmentPacker
-from keras_nlp.models.bert.bert_presets import vocabularies
+from keras_nlp.models.bert.bert_models import Bert
 from keras_nlp.tokenizers.word_piece_tokenizer import WordPieceTokenizer
 
 
-def _handle_pretrained_tokenizer_arguments(vocabulary, lowercase):
-    """Look up pretrained defaults for tokenizer arguments.
+def _download_vocabulary(preset, url, hash):
```
[the remark this answers is not in the document]
Comment (jbischof) [reply keras_nlp/models/bert/bert_preprocessing.py:25]: Good call!

**Comment by jbischof** [comment]
@chenmoneygithub the preprocessing layer does not handle batch inputs. It is expected to be mapped over each example in the batch. If there are multiple sentences in the input, these are interpreted as multiple segments for the same example (e.g., for NSP or sentence similarity).

**Comment by mattdangerw** [comment]
@jbischof @chenmoneygithub I think this batching convo is unrelated to the PR. But seems worth clearing up.

 - All our preprocessing does handle batched and unbatched input. Loading a dataset with, say, tfds batched, and piping it through our preprocessing is actually a pretty core use case! (And we do support it universally)
 - All our preprocessing supports converting to tensors, just as a convenience. (Most tf ops do as well)
 - The `BertPreprocessor` supports multiple (batched or unbatched) inputs which are then collated with `sep` tokens. This is also the behavior of the `MultiSegmentPacker` layer.

The sum total of these means you need to be careful when passing to un-tensorized inputs to the layer, because you hit an ambiguity of sorts. We treat the first tuple or list we see as indicating multiple separate string features to collate. So in the code snippet you posted Chen, we are doing a very valid thing, and contacting the two sentences together in a single sequence. This colab shows what I mean.
https://gist.github.com/mattdangerw/61b74938a11b61f1ce858b770c601bd6

We could consider flipping the default behavior there, so if you pass a tuple of strings (rather than a tuple of numpy or tensors) we treat this as a batch. That is definitely something for a separate issue though!

**Review by mattdangerw — APPROVED** [review]
LGTM! Thanks!
